test/videoframeprocessor/DefaultVideoTransformDevice.test.ts

```diff
@@ -0,0 +1,467 @@
+    processingDidStart = stub();
+    processingDidStop = stub();
+    processingDidFailToStart = stub();
+    processingLatencyTooHigh = stub();
+  }
+
+  beforeEach(() => {
+    domMockBehavior = new DOMMockBehavior();
+    domMockBuilder = new DOMMockBuilder(domMockBehavior);
+    domMockBehavior.createElementCaptureStream = new MediaStream();
+    mockVideoStream = new MediaStream();
+    // @ts-ignore
+    mockVideoStream.id = 'sample';
+    // @ts-ignore
+    mockVideoTrack = new MediaStreamTrack('test', 'video');
+    mockVideoStream.addTrack(mockVideoTrack);
+  });
+
+  afterEach(() => {
+    if (domMockBuilder) {
+      domMockBuilder.cleanup();
+    }
+  });
+
+  describe('construction', () => {
+    it('can be constructed', () => {
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      assert.exists(device);
+    });
+  });
+
+  describe('getter outputMediaStream', () => {
+    it('returns null if device is not started', async () => {
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      expect(await device.outputMediaStream).to.be.null;
+    });
+  });
+
+  describe('applyProcessors', () => {
+    it('set inputMediaStream', async () => {
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      await device.applyProcessors(null);
+      // @ts-ignore
+      expect(device.inputMediaStream).to.be.null;
+    });
+
+    it('does not start if processors is null', async () => {
+      const device = new DefaultVideoTransformDevice(logger, 'test', []);
+      await device.applyProcessors(null);
+    });
+
+    it('returns null if device is not started', done => {
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.applyProcessors(mockVideoStream);
+
+      new TimeoutScheduler(200).start(() => {
+        device.stop();
+      });
+
+      new TimeoutScheduler(50).start(() => {
+        done();
+      });
+    });
+
+    it('can start device', async () => {
+      const obs = new MockObserver();
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      domMockBehavior.createElementCaptureStream = mockVideoStream;
+      await device.applyProcessors(mockVideoStream);
+
+      // a little bit of time to let processing run
+      await new Promise(resolve => setTimeout(resolve, 200));
+
+      expect(obs.processingDidStart.called).to.equal(true);
+      assert.exists(device.outputMediaStream);
+
+      await device.stop();
+      await new Promise(resolve => setTimeout(resolve, 50));
+      expect(obs.processingDidStop.called).to.equal(true);
```

Comment:
This really makes me raise my eyebrows.

Reply:
removed eyebrow-raising timeout scheduler.